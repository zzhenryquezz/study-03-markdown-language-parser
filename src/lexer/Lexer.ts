import type LexerProcessor from './LexerProcessor'
import Token, { TokenType } from './Token'


const files = import.meta.glob('./processors/*.ts', {
    eager: true,
})

const processors = Object.values(files)
    .filter((file: any) => file.default)
    .map((file: any) => new file.default())

export default class Lexer {
    
    public processors: LexerProcessor[] = processors

    public isAlpha(char: string){
        return /[a-zA-Z0-9?]/.test(char)
    }

    public isWhitespace(char: string){
        return /\s/.test(char)
    }

    public isText(char: string){
        return this.isAlpha(char) || this.isWhitespace(char)
    }

    public tokenize(code: string){
        const tokens: Token[] = []

        const chars = code.split('')

        while(chars.length){
            const current = chars[0]

            // console.log('current', current)

            this.processors.sort((a, b) => a.order - b.order)

            const result = this.processors.find((p) => p.process(current, chars, tokens))

            if (result) continue

            console.log('[lexer] unhandled char', current)

            chars.shift()
        }

        tokens.push(Token.from(TokenType.EndOfFile, ''))

        return tokens
    }
}