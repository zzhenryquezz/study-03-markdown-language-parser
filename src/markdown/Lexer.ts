import Token, { TokenType } from './Token'

export default class Lexer {

    public isAlpha(char: string){
        return /[a-zA-Z0-9?]/.test(char)
    }

    public isWhitespace(char: string){
        return /\s/.test(char)
    }

    public isText(char: string){
        return this.isAlpha(char) || this.isWhitespace(char)
    }

    public tokenize(code: string){
        const tokens: Token[] = []

        const chars = code.split('')

        while(chars.length){
            const current = chars[0]

            if(current === '#'){
                tokens.push(Token.from(TokenType.Hashtag, current))

                chars.shift()
                continue
            }

            // check for close tag

            if (current === '<' && chars[1] === '/'){
                const endIndex = chars.findIndex((c) => c === '>') + 1

                const tag = chars.slice(0, endIndex).join('')

                tokens.push(Token.from(TokenType.CloseHTMLTag, tag))

                chars.splice(0, endIndex)

                continue
            }

            // check for open tag

            if (current === '<'){
                const endIndex = chars.findIndex((c) => c === '>') + 1

                const tag = chars.slice(0, endIndex).join('')

                tokens.push(Token.from(TokenType.OpenHTMLTag, tag))

                chars.splice(0, endIndex)

                continue
            }

            // check for whitespace

            if (this.isWhitespace(current)){
                let endIndex = chars.findIndex(c => !this.isWhitespace(c))
                
                if (endIndex === -1) endIndex = chars.length

                const whitespace = chars.slice(0, endIndex).join('')

                tokens.push(Token.from(TokenType.WhiteSpace, whitespace))

                if (endIndex > 0) chars.splice(0, endIndex)
                
                if (endIndex === 0) chars.shift()

                continue
            }

            // check for word
            
            if (this.isAlpha(current)){
                let endIndex = chars.findIndex(c => !this.isAlpha(c))
                
                if (endIndex === -1) endIndex = chars.length

                const word = chars.slice(0, endIndex).join('')

                tokens.push(Token.from(TokenType.Word, word))

                if (endIndex > 0) chars.splice(0, endIndex)
                
                if (endIndex === 0) chars.shift()

                continue
            }

            console.log('unhandled char', current)

            chars.shift()
        }

        return tokens
    }
}