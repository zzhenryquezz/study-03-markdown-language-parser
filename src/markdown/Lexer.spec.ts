import { describe, it, expect } from 'vitest'
import Lexer from './Lexer'
describe('lexer', () => {

    const lexer = new Lexer()
    
    const files = import.meta.glob('../tests/fixtures/*.md', {
        eager: true,
        as: 'raw'
    })

    it.each(Object.entries(files))('should tokenize fixtures %s', (filename, content) => {
        expect(lexer.tokenize(content)).toMatchSnapshot()        
    })
})