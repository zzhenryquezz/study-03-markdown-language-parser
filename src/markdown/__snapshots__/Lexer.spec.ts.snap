// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lexer > should tokenize fixtures ../tests/fixtures/block.code.01.md 1`] = `
[
  Token {
    "type": "Symbol",
    "value": "[",
  },
  Token {
    "type": "Word",
    "value": "code",
  },
  Token {
    "type": "Symbol",
    "value": "]",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "Word",
    "value": "const",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "sum",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Symbol",
    "value": "=",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Symbol",
    "value": "(",
  },
  Token {
    "type": "Word",
    "value": "a",
  },
  Token {
    "type": "Symbol",
    "value": ",",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "b",
  },
  Token {
    "type": "Symbol",
    "value": ")",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Symbol",
    "value": "=",
  },
  Token {
    "type": "Symbol",
    "value": ">",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "a",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Symbol",
    "value": "+",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "b",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "Word",
    "value": "const",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "result",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Symbol",
    "value": "=",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "sum",
  },
  Token {
    "type": "Symbol",
    "value": "(",
  },
  Token {
    "type": "Word",
    "value": "1",
  },
  Token {
    "type": "Symbol",
    "value": ",",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "2",
  },
  Token {
    "type": "Symbol",
    "value": ")",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "BreakLine",
    "value": "
",
  },
  Token {
    "type": "Symbol",
    "value": "[",
  },
  Token {
    "type": "Symbol",
    "value": "/",
  },
  Token {
    "type": "Word",
    "value": "code",
  },
  Token {
    "type": "Symbol",
    "value": "]",
  },
]
`;

exports[`lexer > should tokenize fixtures ../tests/fixtures/heading.01.md 1`] = `
[
  Token {
    "type": "Hashtag",
    "value": "#",
  },
  Token {
    "type": "Hashtag",
    "value": "#",
  },
  Token {
    "type": "Hashtag",
    "value": "#",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "Title",
  },
]
`;

exports[`lexer > should tokenize fixtures ../tests/fixtures/heading.02.md 1`] = `
[
  Token {
    "type": "Hashtag",
    "value": "#",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "Hello",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "123",
  },
]
`;

exports[`lexer > should tokenize fixtures ../tests/fixtures/html.tag.b.01.md 1`] = `
[
  Token {
    "type": "Symbol",
    "value": "<",
  },
  Token {
    "type": "Word",
    "value": "b",
  },
  Token {
    "type": "Symbol",
    "value": ">",
  },
  Token {
    "type": "Word",
    "value": "Hello",
  },
  Token {
    "type": "Symbol",
    "value": "<",
  },
  Token {
    "type": "Symbol",
    "value": "/",
  },
  Token {
    "type": "Word",
    "value": "b",
  },
  Token {
    "type": "Symbol",
    "value": ">",
  },
]
`;

exports[`lexer > should tokenize fixtures ../tests/fixtures/html.tag.b.02.md 1`] = `
[
  Token {
    "type": "Word",
    "value": "Hello",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Symbol",
    "value": "<",
  },
  Token {
    "type": "Word",
    "value": "b",
  },
  Token {
    "type": "Symbol",
    "value": ">",
  },
  Token {
    "type": "Word",
    "value": "Will",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "Smith",
  },
  Token {
    "type": "Symbol",
    "value": "<",
  },
  Token {
    "type": "Symbol",
    "value": "/",
  },
  Token {
    "type": "Word",
    "value": "b",
  },
  Token {
    "type": "Symbol",
    "value": ">",
  },
]
`;

exports[`lexer > should tokenize fixtures ../tests/fixtures/html.tag.b.03.md 1`] = `
[
  Token {
    "type": "Word",
    "value": "Hello",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Symbol",
    "value": "<",
  },
  Token {
    "type": "Word",
    "value": "b",
  },
  Token {
    "type": "Symbol",
    "value": ">",
  },
  Token {
    "type": "Word",
    "value": "Will",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "Smith",
  },
  Token {
    "type": "Symbol",
    "value": "<",
  },
  Token {
    "type": "Symbol",
    "value": "/",
  },
  Token {
    "type": "Word",
    "value": "b",
  },
  Token {
    "type": "Symbol",
    "value": ">",
  },
  Token {
    "type": "Symbol",
    "value": "\\\\",
  },
  Token {
    "type": "Word",
    "value": "n",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "How",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "are",
  },
  Token {
    "type": "WhiteSpace",
    "value": " ",
  },
  Token {
    "type": "Word",
    "value": "you",
  },
  Token {
    "type": "Symbol",
    "value": "?",
  },
]
`;
